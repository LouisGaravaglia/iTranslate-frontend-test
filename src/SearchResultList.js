import React, {useState} from 'react';
import SearchResult from "./SearchResult";
import PaginationSlider from "./PaginationSlider";

const SearchResultList = ({resultsArray, handleSearch, itemsPerPage, typeOfResults, loadingIcon, typeOfTracks, typeOfArtists, typeOfAlbums}) => {
  const [sliderVal, setSliderVal] = useState(0);
  const resultsInView = resultsArray.slice(sliderVal * itemsPerPage, (sliderVal * itemsPerPage) + itemsPerPage);
  const needsPaginationSlider = resultsArray.length > itemsPerPage;
  
  const updateResultsInView = (val) => {
    setSliderVal(val);
  };

////////////////////////////////////////////////////  SEARCH RESULTS  ////////////////////////////////////////////////////

  let displaySearchResults;

  if (typeOfResults === "search-results") displaySearchResults = (
    <>
      <div className="Result-Box">
        {loadingIcon}
        {resultsInView.map((r, i) => <SearchResult key={r.id} index={i} typeOfResults="search-results" handleClick={handleSearch} previewURL={r.preview_url} artist={r.artists[0].name} album={r.album.name} track={r.name} musicObject={r}/>)}
      </div> 
      {needsPaginationSlider && <PaginationSlider resultsArray={resultsArray} itemsPerPage={itemsPerPage} handleSliderChange={updateResultsInView} containerClass="Main-Pagination-Slider-Container" sliderClass="Main-Pagination-Slider"/>}
      {!needsPaginationSlider && <div className="Main-Pagination-Slider-Placeholder"></div>}
    </>
  );

////////////////////////////////////////////////////  DISPLAY ARTISTS  ////////////////////////////////////////////////////

  let displayArtists;

  if (typeOfResults === "artists") displayArtists = (
    <>
      <div className="Result-Box">
        {loadingIcon}
        {resultsInView.map((r, i) => <SearchResult key={i} index={i} typeOfResults="artists" handleClick={handleSearch} artist={r.artistName} spotify_id={r.artistId} typeOfArtists={typeOfArtists} needsPaginationSlider={needsPaginationSlider}/>)}
      </div>
      {needsPaginationSlider && <PaginationSlider  resultsArray={resultsArray} itemsPerPage={itemsPerPage} handleSliderChange={updateResultsInView} containerClass="Main-Pagination-Slider-Container" sliderClass="Main-Pagination-Slider"/>}
    </>
  );

////////////////////////////////////////////////////  DISPLAY ALBUMS  ////////////////////////////////////////////////////

  let displayAlbums;

  if (typeOfResults === "albums") displayAlbums = (
    <>
      <div className="Browse-Albums">
        {resultsInView.map((r, i) => <SearchResult key={i} index={i} typeOfResults="albums" handleClick={handleSearch} name={r.albumName} id={r.albumId} image={r.albumImg} typeOfAlbums={typeOfAlbums}/>)}
      </div>
      {needsPaginationSlider && <PaginationSlider  resultsArray={resultsArray} itemsPerPage={itemsPerPage} handleSliderChange={updateResultsInView} containerClass="Main-Pagination-Slider-Container" sliderClass="Main-Pagination-Slider"/>}
    </>
  );

////////////////////////////////////////////////////  DISPLAY TRACKS  ////////////////////////////////////////////////////

  let displayTracks;

  if (typeOfResults === "tracks") displayTracks = (
    <>
      <div className="Result-Box">
        {resultsInView.map((r, i) => <SearchResult key={r.trackId} index={i} typeOfResults="tracks" handleClick={handleSearch} artistName={r.artistName} trackName={r.trackName} previewURL={r.previewURL} musicObject={r} typeOfTracks={typeOfTracks} needsPaginationSlider={needsPaginationSlider}/>)}
      </div>
      {needsPaginationSlider && <PaginationSlider  resultsArray={resultsArray} itemsPerPage={itemsPerPage} handleSliderChange={updateResultsInView} containerClass="Main-Pagination-Slider-Container" sliderClass="Main-Pagination-Slider"/>}
    </>
  );


////////////////////////////////////////////////////  DISPLAY GENRES  ////////////////////////////////////////////////////

  let displayGenres;

  if (typeOfResults === "genres") displayGenres = (
    <>
      <div className="Result-Box">
        {resultsInView.map((r, i) => <SearchResult key={i} index={i} typeOfResults="genres" handleClick={handleSearch} genre={r}/>)}
      </div> 
      {needsPaginationSlider && <PaginationSlider  resultsArray={resultsArray} itemsPerPage={itemsPerPage} handleSliderChange={updateResultsInView} containerClass="Main-Pagination-Slider-Container" sliderClass="Main-Pagination-Slider"/>}
      {!needsPaginationSlider && <div className="Main-Pagination-Slider-Placeholder"></div>}
    </>
  );

////////////////////////////////////////////////////  DANCEABILITY RESULTS  ////////////////////////////////////////////////////

  let displayDanceabilityTracks;

  if (typeOfResults === "danceability-results") displayDanceabilityTracks = (
    <>
      <div className="Danceability-Result-Box">
        {resultsArray.length && resultsInView.map((r, i) => <SearchResult key={r.trackId} index={i} typeOfResults="danceability-results" previewURL={r.previewURL} handleClick={handleSearch} artist={r.artistName} album={r.albumName} track={r.trackName} musicObject={r} typeOfTracks="danceability"/>)}
      </div>
      {needsPaginationSlider && <PaginationSlider  resultsArray={resultsArray} itemsPerPage={itemsPerPage} handleSliderChange={updateResultsInView} containerClass="Danceability-Pagination-Slider-Container" sliderClass="Danceability-Pagination-Slider"/>}
      {!needsPaginationSlider && <div className="Temp-v2"></div>}
    </>
  );

////////////////////////////////////////////////////  RETURN  ////////////////////////////////////////////////////

  return (
    <>
      {displaySearchResults}
      {displayArtists}
      {displayAlbums}
      {displayTracks}
      {displayGenres}
      {displayDanceabilityTracks}
    </>
  );
};

export default SearchResultList;