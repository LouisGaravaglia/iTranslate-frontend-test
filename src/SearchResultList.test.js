import React from 'react'
import {render, cleanup} from '@testing-library/react'
import '@testing-library/jest-dom/extend-expect'
import {Provider} from 'react-redux'
import {createStore} from "redux";
import SearchResultList from "./SearchResultList";

afterEach(cleanup);

const startingState = {
  allArtists: [{trackName: "Frank Ocean"}],
  selectedTrack: {trackName: "chanel", trackId: 42}, 
  errors: {translationError: false, languageError: false, lyricsError: false, searchError: false, generalError: false}
  };

/**
 * reducer function to mimic the reducer used in Redux.
 * @param {object} state - holds an object of data used within the components
 * @param {object} action - object where the values are strings used to fire certain actions
 */
function reducer(state = startingState, action) {
  switch (action.type) {
    case "ADD_TRACKS":
      return {...state, tracks: action.tracks};
    default:
      return state;
  };
};

/**
 * Function that renders the component within a redux environment since most 
 * components rely on redux for data in order to mount.
 * @param {ReactComponent} component - private spotify client id
 * @param {function} store - creates the react store using the createStore method imported from redux
 */
function renderWithRedux(
  component,
  {initialState, store = createStore(reducer, initialState)} = {}
) {
  return {
    ...render(<Provider store={store}>{component}</Provider>)
  };
};

//SMOKE TEST
describe('Smoke Test for SearchResultList component', () => {

  it('renders without crashing', () => {
    renderWithRedux(<SearchResultList itemsPerPage={1} resultsArray={[{trackId: 1}, {trackId: 2}]}/>);
  });
});

//SNAPSHOT TEST
describe('Snapshot Test for SearchResultList component', () => {

  it('matches snapshot', () => {
    const {asFragment} = renderWithRedux(<SearchResultList resultsArray={[{trackId: 1}, {trackId: 2} ]} itemsPerPage={2}/>);
    expect(asFragment()).toMatchSnapshot();
  });
});