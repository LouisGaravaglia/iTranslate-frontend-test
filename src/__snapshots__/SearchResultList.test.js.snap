// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot Test for SearchResultList component matches snapshot 1`] = `<DocumentFragment />`;
